From 849ea866b7b7c1c92e313483b7782e4aa1feaf29 Mon Sep 17 00:00:00 2001
From: Qiang Yu <yuq825@gmail.com>
Date: Thu, 6 Dec 2018 20:00:44 +0800
Subject: [PATCH 45/46] drm/lima: return error status of fence after waiting

For user space driver know which draw has error in
kernel space.

Signed-off-by: Qiang Yu <yuq825@gmail.com>
---
 drivers/gpu/drm/lima/lima_drv.c | 3 +++
 include/uapi/drm/lima_drm.h     | 4 ++--
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/lima/lima_drv.c b/drivers/gpu/drm/lima/lima_drv.c
index 536d0d4..4f0060e 100644
--- a/drivers/gpu/drm/lima/lima_drv.c
+++ b/drivers/gpu/drm/lima/lima_drv.c
@@ -213,8 +213,11 @@ static int lima_ioctl_wait_fence(struct drm_device *dev, void *data, struct drm_
 
 	if (fence) {
 		err = lima_wait_fence(fence, args->timeout_ns);
+		args->error = fence->error;
 		dma_fence_put(fence);
 	}
+	else
+		args->error = 0;
 
 	return err;
 }
diff --git a/include/uapi/drm/lima_drm.h b/include/uapi/drm/lima_drm.h
index 0c17bf6..c44757b 100644
--- a/include/uapi/drm/lima_drm.h
+++ b/include/uapi/drm/lima_drm.h
@@ -137,7 +137,7 @@ struct drm_lima_wait_fence {
 	__u32 pipe;        /* in */
 	__u64 timeout_ns;  /* in */
 	__u32 seq;         /* in */
-	__u32 _pad;
+	__u32 error;       /* out */
 };
 
 #define LIMA_GEM_WAIT_READ   0x01
@@ -181,7 +181,7 @@ struct drm_lima_gem_mod {
 #define DRM_IOCTL_LIMA_GEM_INFO DRM_IOWR(DRM_COMMAND_BASE + DRM_LIMA_GEM_INFO, struct drm_lima_gem_info)
 #define DRM_IOCTL_LIMA_GEM_VA DRM_IOW(DRM_COMMAND_BASE + DRM_LIMA_GEM_VA, struct drm_lima_gem_va)
 #define DRM_IOCTL_LIMA_GEM_SUBMIT DRM_IOWR(DRM_COMMAND_BASE + DRM_LIMA_GEM_SUBMIT, union drm_lima_gem_submit)
-#define DRM_IOCTL_LIMA_WAIT_FENCE DRM_IOW(DRM_COMMAND_BASE + DRM_LIMA_WAIT_FENCE, struct drm_lima_wait_fence)
+#define DRM_IOCTL_LIMA_WAIT_FENCE DRM_IOWR(DRM_COMMAND_BASE + DRM_LIMA_WAIT_FENCE, struct drm_lima_wait_fence)
 #define DRM_IOCTL_LIMA_GEM_WAIT DRM_IOW(DRM_COMMAND_BASE + DRM_LIMA_GEM_WAIT, struct drm_lima_gem_wait)
 #define DRM_IOCTL_LIMA_CTX DRM_IOWR(DRM_COMMAND_BASE + DRM_LIMA_CTX, struct drm_lima_ctx)
 #define DRM_IOCTL_LIMA_GEM_MOD DRM_IOWR(DRM_COMMAND_BASE + DRM_LIMA_GEM_MOD, struct drm_lima_gem_mod)
-- 
2.7.4

